src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -746,8 +753,16 @@ private void EnforceDoesNotReturn(SyntaxNode? syntaxOpt)
             DiagnosticBag diagnostics,
             bool createSnapshots,
             out SnapshotManager? snapshotManager,
-            ref ImmutableDictionary<Symbol, Symbol>? remappedSymbols)
+            ref ImmutableDictionary<Symbol, Symbol>? remappedSymbols,
+            bool avoidRewrite)
         {
+            // If we're avoiding a rewrite, then we must be in debug mode, and semantic analysis must be globally disabled.
+#if DEBUG
+            Debug.Assert(avoidRewrite != compilation.NullableSemanticAnalysisEnabled);
+#else
+            if (avoidRewrite) throw ExceptionUtilities.Unreachable;
```

Comment:
Overall this dance feels a bit awkward. We're using runtime checks to enforce behavior that is actually codified at compile time. I think this would be a lot cleaner if we split it up into two methods: `Analyze` and `AnalyzeAndRewrite`.